### PR TITLE
Feature/dhfprod 2076 step definition

### PR DESCRIPTION
**Pull request by akshaysonvane**
Add AbstractStepDefinition model with concrete implementations
Convert Step model to simple POJO
Adjust gradle tasks and tests

Note: Basic functionality is all working now
TODO: manually tests for all scenarios and optimize code

### CODE REVIEW

File: marklogic-data-hub/src/main/java/com/marklogic/hub/step/StepDefinition.java [hunk 1]
```diff
@@ -17,13 +17,13 @@
...
+import com.marklogic.hub.step.impl.*;
 
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Map;
 
-public interface Step {
+public interface StepDefinition {
 
     enum StepType {
```
Comment (srinathgit) [inline marklogic-data-hub/src/main/java/com/marklogic/hub/step/StepDefinition.java]: Should the enum be called StepDefinitionType ?

File: marklogic-data-hub/src/main/java/com/marklogic/hub/step/StepDefinition.java [hunk 2]
```diff
@@ -62,8 +62,25 @@ public String toString() {
...
      * @return a Step object
      */
-    static Step create(String name, StepType type) {
-        return new StepImpl(name, type);
+    static StepDefinition create(String name, StepType type) {
+        StepDefinition stepDefinition = null;
+
+        switch (type) {
+            case INGEST:
+                stepDefinition = new IngestionStepDefinitionImpl(name);
```
Comment (srinathgit) [inline marklogic-data-hub/src/main/java/com/marklogic/hub/step/StepDefinition.java]: Didn't this change to "INGESTION" and "MASTERING" ?

File: web/src/main/java/com/marklogic/hub/web/service/FlowManagerService.java [hunk 3]
```diff
@@ -137,29 +138,36 @@ public StepModel createStep(String flowName, Integer stepOrder, String stepId, S
...
+        if (step.getStepDefinitionType() == null) {
             throw new BadRequestException("Invalid Step Type");
         }
 
         // Only save step if step is of Custom type, for rest use the default steps.
-        switch (step.getType()) {
+        switch (step.getStepDefinitionType()) {
             case INGEST:
-                Step defaultIngest = getDefaultStepFromResources("hub-internal-artifacts/steps/ingest/marklogic/default-ingest.step.json", Step.StepType.INGEST);
+                StepDefinition defaultIngestDefinition = getDefaultStepDefinitionFromResources("hub-internal-artifacts/steps/ingest/marklogic/default-ingest.step.json", StepDefinition.StepType.INGEST);
```
Comment (srinathgit) [inline web/src/main/java/com/marklogic/hub/web/service/FlowManagerService.java]: should be "INGESTION"

File: web/src/main/java/com/marklogic/hub/web/service/FlowManagerService.java [hunk 4]
```diff
@@ -137,29 +138,36 @@ public StepModel createStep(String flowName, Integer stepOrder, String stepId, S
...
                 break;
             case MAPPING:
-                Step defaultMapping = getDefaultStepFromResources("hub-internal-artifacts/steps/mapping/marklogic/default-mapping.step.json", Step.StepType.MAPPING);
+                StepDefinition defaultMappingDefinition = getDefaultStepDefinitionFromResources("hub-internal-artifacts/steps/mapping/marklogic/default-mapping.step.json", StepDefinition.StepType.MAPPING);
+                Step defaultMapping = defaultMappingDefinition.transformToStep(step.getName(), defaultMappingDefinition, new Step());
                 step = StepModel.mergeFields(stepModel, defaultMapping, step);
                 break;
             case MASTER:
-                Step defaultMaster = getDefaultStepFromResources("hub-internal-artifacts/steps/master/marklogic/default-mastering.step.json", Step.StepType.MASTER);
+                StepDefinition defaultMasterDefinition = getDefaultStepDefinitionFromResources("hub-internal-artifacts/steps/master/marklogic/default-mastering.step.json", StepDefinition.StepType.MASTER);
```
Comment (srinathgit) [inline web/src/main/java/com/marklogic/hub/web/service/FlowManagerService.java]: should be "MASTERING"

File: marklogic-data-hub/src/main/java/com/marklogic/hub/StepDefinitionManager.java [hunk 5]
```diff
@@ -64,29 +64,29 @@
      * @param type - type of the Step
      * @return the Step object
      */
```
Comment (srinathgit) [inline marklogic-data-hub/src/main/java/com/marklogic/hub/StepDefinitionManager.java]: java docs have to be updated in all the places

File: marklogic-data-hub/src/main/java/com/marklogic/hub/step/StepDefinition.java (hunk 2 above)
Comment (akshaysonvane) [reply marklogic-data-hub/src/main/java/com/marklogic/hub/step/StepDefinition.java]: Didn't want to break the current code.
Should be taken up as part of https://project.marklogic.com/jira/browse/DHFPROD-2096

File: web/src/main/java/com/marklogic/hub/web/service/FlowManagerService.java (hunk 3 above)
Comment (akshaysonvane) [reply web/src/main/java/com/marklogic/hub/web/service/FlowManagerService.java]: Didn't want to break the current code.
Should be taken up as part of https://project.marklogic.com/jira/browse/DHFPROD-2096

File: web/src/main/java/com/marklogic/hub/web/service/FlowManagerService.java (hunk 4 above)
Comment (akshaysonvane) [reply web/src/main/java/com/marklogic/hub/web/service/FlowManagerService.java]: Didn't want to break the current code.
Should be taken up as part of https://project.marklogic.com/jira/browse/DHFPROD-2096

File: marklogic-data-hub/src/main/java/com/marklogic/hub/step/StepDefinition.java (hunk 1 above)
Comment (akshaysonvane) [reply marklogic-data-hub/src/main/java/com/marklogic/hub/step/StepDefinition.java]: Yep missed that.
Done https://github.com/marklogic/marklogic-data-hub/pull/2207/commits/168804fa9b372d529eb83717f74637bdc401254e

File: marklogic-data-hub/src/main/java/com/marklogic/hub/step/impl/IngestionStepDefinitionImpl.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    public IngestionStepDefinitionImpl(String name) {
+        setName(name);
+        setType(StepDefinitionType.INGEST);
+        setFileLocations(fileLocations);
+
+        Map<String, Object> options = getOptions();
+        options.put("outputFormat", "json");
+
+        setModulePath("/data-hub/5/builtins/steps/ingest/default/main.sjs");
+    }
```
Comment (srinathgit) [inline marklogic-data-hub/src/main/java/com/marklogic/hub/step/impl/IngestionStepDefinitionImpl.java]: Should this module path be changed and similarly for mastering, mapping etc ?

File: marklogic-data-hub/src/main/java/com/marklogic/hub/deploy/commands/LoadUserArtifactsCommand.java [hunk 7]
```diff
@@ -137,7 +137,7 @@ public String toURI(Resource r) throws IOException {
         };
         ResourceToURI stepResourceToURI = new ResourceToURI(){
             public String toURI(Resource r) throws IOException {
-                return "/steps/" + r.getFile().getParentFile().getParentFile().getName() + "/" + r.getFile().getParentFile().getName() + "/" + r.getFilename();
+                return "/step-definitions/" + r.getFile().getParentFile().getParentFile().getName() + "/" + r.getFile().getParentFile().getName() + "/" + r.getFilename();
```
Comment (srinathgit) [inline marklogic-data-hub/src/main/java/com/marklogic/hub/deploy/commands/LoadUserArtifactsCommand.java]: Should the collection name change to http://marklogic.com/data-hub/step-definition ? If so, there are quite a few places in enode code that has to be changed as well.

Comment (akshaysonvane) [reply marklogic-data-hub/src/main/java/com/marklogic/hub/deploy/commands/LoadUserArtifactsCommand.java]: I've added 
`// TODO: Rename collection URI after changing server code`
for that change as I dont want to break the enode code.
Should be taken up as part of https://project.marklogic.com/jira/browse/DHFPROD-2096

File: marklogic-data-hub/src/main/java/com/marklogic/hub/step/impl/IngestionStepDefinitionImpl.java (hunk 6 above)
Comment (akshaysonvane) [reply marklogic-data-hub/src/main/java/com/marklogic/hub/step/impl/IngestionStepDefinitionImpl.java]: Yes, I've added 
`// TODO: Rename internal artifacts step folder to step-definitions, change collections URI` 
for that change as I dont want to break the enode code.
Should be taken up as part of https://project.marklogic.com/jira/browse/DHFPROD-2096

File: marklogic-data-hub/src/main/java/com/marklogic/hub/deploy/commands/LoadUserArtifactsCommand.java (hunk 7 above)
Comment (srinathgit) [reply marklogic-data-hub/src/main/java/com/marklogic/hub/deploy/commands/LoadUserArtifactsCommand.java]: But is the change from step to step-definition part of the scope of https://project.marklogic.com/jira/browse/DHFPROD-2096 ? I thought DHFPROD-2076 handles that .

Comment (akshaysonvane) [reply marklogic-data-hub/src/main/java/com/marklogic/hub/deploy/commands/LoadUserArtifactsCommand.java]: Can be taken up as part of DHFPROD-2076 as well, but I don't want to break the code for now.

**Pull request merged**